kernel/frame_allocator/src/lib.rs

```diff
@@ -263,112 +322,291 @@ pub enum MemoryRegionType {
 }
 
 /// A range of contiguous frames.
+/// Owning a `Frames` object gives ownership of the range of frames it references.
+/// 
+/// The frames can be in a free, allocated, mapped or unmapped state. 
+/// In the free state, frames are owned by the frame allocator and have not been allocated for a mapping.
+/// In the allocated state, frames have been removed from the free list and can be used for a mapping.
+/// In the mapped state, frames have been mapped to a virtual memory page and are in use.
+/// In the unmapped state, frames have been unmapped and can be retured to the frame allocator.
+/// 
+/// When a `Frames` object in a free state is dropped, it will be added back to the free list.
+/// When a `Frames` object in an allocated state is dropped, it is transitioned to a free state. 
+/// When a `Frames` obect in an unmapped state is dropped, it is transitioned to an allocated state.
+/// We expect that `Frames` in a mapped state will never be dropped, but instead will be forgotten and then 
+/// recreated in an unmapped state when it's frames are cleared from the page table.
 ///
+///     (Free) <--> (Allocated) -> (Mapped) -> (Unmapped) -> (Allocated)
+/// 
 /// # Ordering and Equality
 ///
-/// `Chunk` implements the `Ord` trait, and its total ordering is ONLY based on
-/// its **starting** `Frame`. This is useful so we can store `Chunk`s in a sorted collection.
+/// `Frames` implements the `Ord` trait, and its total ordering is ONLY based on
+/// its **starting** `Frame`. This is useful so we can store `Frames` in a sorted collection.
 ///
-/// Similarly, `Chunk` implements equality traits, `Eq` and `PartialEq`,
-/// both of which are also based ONLY on the **starting** `Frame` of the `Chunk`.
-/// Thus, comparing two `Chunk`s with the `==` or `!=` operators may not work as expected.
+/// Similarly, `Frames` implements equality traits, `Eq` and `PartialEq`,
+/// both of which are also based ONLY on the **starting** `Frame` of the `Frames`.
+/// Thus, comparing two `Frames` with the `==` or `!=` operators may not work as expected.
 /// since it ignores their actual range of frames.
-#[derive(Debug, Clone, Eq)]
-struct Chunk {
+#[derive(Eq)]
+pub struct Frames<const S: MemoryState> {
     /// The type of this memory chunk, e.g., whether it's in a free or reserved region.
     typ: MemoryRegionType,
-    /// The Frames covered by this chunk, an inclusive range. 
-    frames: FrameRange,
+    /// The Frames covered by this chunk, an inclusive range.
+    frames: FrameRange
 }
-impl Chunk {
-    fn as_allocated_frames(&self) -> AllocatedFrames {
-        AllocatedFrames {
-            frames: self.frames.clone(),
+
+// Type aliases for the Frames in each state
+pub type FreeFrames = Frames<{MemoryState::Free}>;
+pub type AllocatedFrames = Frames<{MemoryState::Allocated}>;
+pub type MappedFrames = Frames<{MemoryState::Mapped}>;
+pub type UnmappedFrames = Frames<{MemoryState::Unmapped}>;
```

Comment:
for correct docs appearance, we should use `///` doc comments for each type alias.

```suggestion
/// A type alias for `Frames` in the `Free` state.
pub type FreeFrames = Frames<{MemoryState::Free}>;
/// A type alias for `Frames` in the `Allocated` state.
pub type AllocatedFrames = Frames<{MemoryState::Allocated}>;
/// A type alias for `Frames` in the `Mapped` state.
pub type MappedFrames = Frames<{MemoryState::Mapped}>;
/// A type alias for `Frames` in the `Unmapped` state.
pub type UnmappedFrames = Frames<{MemoryState::Unmapped}>;
```